src/engine/input/RunOnInputAttribute.cs

```diff
@@ -0,0 +1,18 @@
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+
+[AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+public abstract class RunOnInputAttribute : Attribute
+{
+    ~RunOnInputAttribute()
+    {
+        AttributesWithMethods.RemoveAll(p => p.Item2.Equals(this));
+    }
+
+    public static IList<object> InputClasses { get; } = new List<object>();
+    public abstract IInputReceiver InputReceiver { get; }
+
+    internal static List<Tuple<MethodBase, RunOnInputAttribute>> AttributesWithMethods { get; }
```

Comment:
Use the named tuple here.
It's a bit weird how you've designed this so that the InputManager is responsible for building this, *but* this class removes stuff from here? I think you should try quite hard to make this responsibility of only one of these classes and not both.